titus_isolate/isolate/cpu.py

```diff
@@ -4,157 +4,174 @@
-    if len(set(curr_ids_per_workload.keys()) ^ set(workload_insertion_times)) != 0:
-        raise Exception("Invalid workload_insertion_times passed: `%s`" % (workload_insertion_times,))
+        if workload_id not in curr_ids_per_workload:
+            raise Exception("workload_id=`%s` is not placed on the instance. Cannot free it." % (workload_id,))
 
-    ordered_workload_ids = [t[0] for t in sorted(workload_insertion_times.items(), key=lambda t: t[1])]
-    tid_2order = {i: t.get_id() for i, t in enumerate(cpu.get_threads())}
+        ordered_workload_ids = self.__ordered_workload_ids()
+        tid_2order = {i: t.get_id() for i, t in enumerate(self.__cpu.get_threads())}
 
-    curr_placement_vectors = []
-    for wid in ordered_workload_ids:
-        cids = curr_ids_per_workload[wid]
-        v = [1 if tid_2order[i] in cids else 0 for i in range(n_compute_units)]
-        curr_placement_vectors.append(v)
-    if len(curr_placement_vectors) == 0:
-        raise Exception("Cannot free a workload from an empty CPU")
+        curr_placement_vectors = []
+        for wid in ordered_workload_ids:
+            cids = curr_ids_per_workload[wid]
+            v = [1 if tid_2order[i] in cids else 0 for i in range(n_compute_units)]
+            curr_placement_vectors.append(v)
+        if len(curr_placement_vectors) == 0:
+            raise Exception("Cannot free a workload from an empty CPU")
 
-    requested_cus = [len(curr_ids_per_workload[wid]) if wid != workload_id else 0 for wid in ordered_workload_ids]
+        requested_cus = [len(curr_ids_per_workload[wid]) if wid != workload_id else 0 for wid in ordered_workload_ids]
 
-    new_placement_vectors = optimize_ip(requested_cus, n_compute_units, n_packages, curr_placement_vectors, verbose=True)
+        new_placement_vectors = self.__compute_new_placement(curr_placement_vectors, requested_cus)
 
-    thread_id2workload_id = {}
-    for w_ind, v in enumerate(new_placement_vectors):
-        if ordered_workload_ids[w_ind] == workload_id:
-            continue
-        for i, e in enumerate(v):
-            if e == 1:
-                thread_id2workload_id[tid_2order[i]] = ordered_workload_ids[w_ind]
+        thread_id2workload_id = {}
+        for w_ind, v in enumerate(new_placement_vectors):
+            if ordered_workload_ids[w_ind] == workload_id:
+                continue
+            for i, e in enumerate(v):
+                if e == 1:
+                    thread_id2workload_id[tid_2order[i]] = ordered_workload_ids[w_ind]
 
-    cpu.clear()
+        self.__assign_new_mapping(thread_id2workload_id)
+        self.__workload_insertion_times.pop(workload_id)
 
-    new_used_threads = []
-    for t in cpu.get_threads():
-        wid = thread_id2workload_id.get(t.get_id(), None)
-        if wid is not None:
-            t.claim(wid)
-            new_used_threads.append(t)
 
-    return new_used_threads
+class GreedyCpuAllocator:
```

Comment:
So far this project avoids defining multiple classes per file except in test scenarios.  It might be nice to maintain that convention.

Reply:
done